docs/ref/templates/builtins.txt

```diff
@@ -953,6 +953,63 @@ output (as a string) inside a variable. This is useful if you want to use
     {% now "Y" as current_year %}
     {% blocktranslate %}Copyright {{ current_year }}{% endblocktranslate %}
 
+.. templatetag:: query_string
+
+``query_string``
+----------------
+
+.. versionadded:: 5.1
+
+Outputs the query string from the given :class:`~django.http.QueryDict`
+instance. If the ``QueryDict`` is empty, then the output will be an empty
+string. Otherwise, the query string will be displayed with a leading "?".
+
+The following example outputs the current query string verbatim. So if the
+query string is ``?color=green&=size=M``, the output would be
+``?color=green&size=M``:
+
+.. code-block:: html+django
+
+    {% query_string request.GET %}
```

Comment:
We want to implement a new builtin template tag so it should work in Django flow, so we don't need to support custom `QueryDict` which is rather niche. That's why I'd prefer to pass `request` instead of `request.GET` or even avoid passing `request` at all and implement it via a `Node` subclass where `GET` is available in `render()` (`context.request.GET`).

Reply:
Hmm, if we don't want to allow a custom QueryDict, then I think it doesn't make much sense to need to pass the request. So I think the second option makes most sense. And I was really hoping to use `simple_tag`, oh well.

Reply:
> And I was really hoping to use simple_tag, oh well.

That's my proposition to make the usage simpler. You can always disagree :disguised_face:

Reply:
No, I think you're right. I tried this approach and realised (unless I'm missing something) it still needs the request context processor to access the request... which means I can again use `simple_tag` 🎉. I added the query_dict as an optional argument in case someone still wants to use it with a custom QueryDict and added a test for this. It's just one extra branch but I can still remove it if you think it's unnecessary. I do agree it's quite niche for someone to want to do this, and either not want to enable the request context processor or want to use a custom QueryDict...

Anyway, I pushed this. I also realised I didn't keep alphabetical ordering in the module so I fixed that as well, though it probably messes up the commit diffs a bit.

Reply:
While not supporting custom `QueryDict`s makes it possible to simplify the tag, I'm not sure if custom `QueryDict`s are really so niche:

In my project, I use URLs like `?year=2023&month=12`. Then in the template for the December report I build links to the previous and next month (e. g. `?year=2024&month=1`). As this is not easy to achieve in the template alone, I have the view precompute a custom `QueryDict`, pass it via the template context and then add or modify extra parameters using the `query_string` tag. For example:
```
{# Explicit, custom QueryDict #}
{% query_string precomputed_querydict_for_next_period format=pdf more_extras=abc %}
```

If the `request.GET` was implied, the same could be achieved alternatively like:
```
{# Implied request.GET #}
{% query_string year=next_period_year month=next_period_month format=pdf more_extras=abc %}
```
but I found this a bit long-winded.

This is why in my original proposal supported multiple dictionaries. It allows something like this:
```
{# Multiple dictionaties. `overrides_for_next_period` is a precomputed dict with keys for year and month #}
{% query_string request.GET overrides_for_next_period format=pdf more_extras=abc %}
```

I may be missing something, but so far I have not found a better (short and clear) way to implement this and thus propose to support this use case. 😜

Reply:
> While not supporting custom `QueryDict`s makes it possible to simplify the tag, I'm not sure if custom `QueryDict`s are really so niche:

I'd argue that a template is not the right place for such a complex mechanism of building query strings.

```
{% query_string year=next_period_year month=next_period_month format=pdf more_extras=abc %}
```
This looks fine to me, TBH.

Reply:
> > While not supporting custom `QueryDict`s makes it possible to simplify the tag, I'm not sure if custom `QueryDict`s are really so niche:
> 
> I'd argue that a template is not the right place for such a complex mechanism of building query strings.

Yes, this is why I found the compromise quite appealing: Have the view precompute query parameters that are complex and “view-centric” while allowing the template to refine the details, e. g. other file formats of the view's data, next page, etc.

> ```
> {% query_string year=next_period_year month=next_period_month format=pdf more_extras=abc %}
> ```
> 
> This looks fine to me, TBH.

Yes, it works, too, especially when the `next_period_*` parameters are grouped like this:

```
{% query_string year=next_period.year month=next_period.month format=pdf more_extras=abc %}
```

Hmmm. Would it be feasible if both the `request.GET` was implied *and* passing extra dicts via
```python
@register.simple_tag
def query_string(*args, **kwargs):
    ...
```
was allowed?

Reply:
It's certainly feasible and the extra complexity is low but I'll leave it to the fellows if they think it's worth doing or not.

For me I think it would see limited use and might make the tag a bit harder to understand as the signature would be more complex.

Reply:
> It's certainly feasible and the extra complexity is low but I'll leave it to the fellows if they think it's worth doing or not.
> 
> For me I think it would see limited use and might make the tag a bit harder to understand as the signature would be more complex.

I agree. Just let me mention one more time that also accepting `dict`s would allow the view to pass parameters to the query string that can be left unnamed in the template.

Reply:
> It's certainly feasible and the extra complexity is low but I'll leave it to the fellows if they think it's worth doing or not.

Maybe the extra complexity is low, but, IMO, it makes API complicated and confusing. Previously it was straightforward , we pass keyword argument that update the current query string. Now it's,
- if you pass a querydict it will be updated with keyword arguments,
- if you don't pass a querydict the current query string is updated with keyword arguments.

For example, Why doesn't passing a querydict update the current query string? (Personally, I'd assume this behavior.) I believe that we should keep this API as simple and straightforward as possible, because templates are not a good place for such a complicated manipulations.

Reply:
I think that your API description is needlessly complex: It updates the items of `request.GET` and returns them as a query string. If a `QueryDict` is given, it uses that instead of `request.GET`.

More importantly, I also think that linking to the next and previous period from month-year report is only marginally more complicated than linking to the next and previous page of a list report. While the latter dominates many examples and there is built-in support in Django for pagination, I believe that linking to a next and previous period of time is a frequent case as well. My suggestion to also support `dict`s as parameters to the `query_string` tag would support that use case, while not supporting it makes the tag useless to me.

My original implementation just took a list of `QueryDict`s and `dict`s, combined them and then modified single items via named parameters. Generic, explicit and still simple, both in API documentation and code. I much like @knyghty 's resulting implementation, but if I could wish, why not just combine explicitly given `dict`s?

Reply:
So, it seems we need to go back to the Forum where we'll reach a wider audience and see what other think, as we will not reach a consensus on use and implementation here.

We have two options:

1. template tag `{% query_string %}` to modify and return the current query string (from `request.GET`) by passing keyword arguments.
2. template tag `{% query_string %}` to modify any query dict by passing keyword argument, other dicts, or query dicts.

I'm against the second option as it's unnecessarily complicated, it seems rather niche, and IMO template tags are not the place for such complex modifications.

Reply:
https://forum.djangoproject.com/t/adding-a-template-tag-to-generate-query-strings/24521/6